File: src-f/test_lapack_2.f90
program test_2
    implicit none
    integer, parameter :: N=1000
    double precision, allocatable :: a(:,:), b(:)
    integer, allocatable :: ipiv(:), jpiv(:)
    double precision :: scal, maxerr
    integer :: i, info
    
    allocate(a(1:N,1:N))
    allocate(b(1:N))
    allocate(ipiv(1:N))
    allocate(jpiv(1:N))
    
    ! matrix from http://www.cs.yale.edu/homes/spielman/BAP/lect6.pdf
    a(1:N,1:N)=0.0
    do i=1, N-1
        a(i,i)=1.0
        a(i+1:N,i)=-1.0
    enddo
    a(1:N,N)=1.0
    do i=1, N
        b(i)=sum(a(i,1:N))
    enddo
    call dgetc2(N,a,N,ipiv,jpiv,info)
    if(info>0)then
        write(*,'(a,i4,a,i4,a,e13.6)') "A(",info,",",info,")=",a(info,info)
        stop 1
    endif
    call dgesc2(N,a,N,b,ipiv,jpiv,scal)
    if(scal/=1.0)then
        b(1:N)=b(1:N)*scal
    endif
    maxerr=0.0
    do i=1, N
        if(abs(b(i)-1.0)>abs(maxerr))then
            maxerr=b(i)-1.0
        endif
    enddo
    write(*,'(e13.6)') maxerr
    if(abs(maxerr)>1.0e-8)then
        write(0,'(A)') "Large error"
        stop 1
    endif
    deallocate(a)
    deallocate(b)
    deallocate(ipiv)
    deallocate(jpiv)
end

File: src-c/test_lapack_2.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
//#include<cblas.h>
//#include<clapack.h>

extern void dgetc2_(const int *, double *, const int *, int *, int *, int *);
extern void dgesc2_(const int *, double *, const int *, double *, int *, int *, double *);

int main()
{
    const int N=1000;
    double *a;
    double *b;
    double scale;
    int *ipiv, *jpiv;
    double maxerr;
    int i,j, info;

    a=malloc(sizeof(double)*N*N);
    if(a==NULL){
	perror("");
	exit(1);
    }
    b=malloc(sizeof(double)*N);
    if(b==NULL){
	perror("");
	exit(1);
    }
    ipiv=malloc(sizeof(int)*N);
    if(ipiv==NULL){
	perror("");
	exit(1);
    }
    jpiv=malloc(sizeof(int)*N);
    if(jpiv==NULL){
	perror("");
	exit(1);
    }
    /* matrix from http://www.cs.yale.edu/homes/spielman/BAP/lect6.pdf */
    for(i=0; i<N; i++){
	for(j=0; j<i; j++){
	    a[i+j*N]=-1.0;
	}
	a[i+i*N]=1.0;
	for(j=i+1; j<N-1; j++){
	    a[i+j*N]=0.0;
	}
	a[i+(N-1)*N]=1.0;
	b[i]=0.0;
	for(j=0; j<N; j++){
	    b[i]+=a[i+j*N];
	}
	ipiv[i]=0;
	jpiv[i]=0;
    }
    /* solve */
    dgetc2_(&N, a, &N, ipiv, jpiv, &info);
    dgesc2_(&N, a, &N, b, ipiv, jpiv, &scale);
    if(scale!=1.0){
	for(i=0; i<N; i++){
	    b[i]*=scale;
	}
    }
    /* output */
    maxerr=0.0;
    for(i=0; i<N; i++){
	if(fabs(b[i]-1.0)>fabs(maxerr)){
	    maxerr=b[i]-1.0;
	}
    }
    printf("%g\n",maxerr);
    if(fabs(maxerr)>1.0e-8){
	fprintf(stderr, "Large error\n");
	exit(1);
    }
    free(a);
    free(b);
    free(ipiv);
    free(jpiv);
    return 0;
}
